unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
     }
 
     for (int i = 0; i <= bytes.length - b.length; i++) {
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final UTF8String prefix, int offset) {
+    byte[] b = prefix.getBytes();
+    if (b.length + offset > bytes.length || offset < 0) {
+      return false;
+    }
+    int i = 0;
+    while (i < b.length && b[i] == bytes[i + offset]) {
+      i++;
+    }
+    return i == b.length;
+  }
+
   public boolean startsWith(final UTF8String prefix) {
-    final byte[] b = prefix.getBytes();
-    // TODO: Avoid copying.
-    return b.length <= bytes.length && Arrays.equals(Arrays.copyOfRange(bytes, 0, b.length), b);
+    return startsWith(prefix, 0);
```

Comment:
Right now, UTF8String is only used by catalyst, can it will check prefix is null or not. Since this is a public method, it's still good to check prefix is null or not.